[JsonSchema] Use arrayObject for objects

Rid opened the pull request:
In order to fix https://github.com/janephp/janephp/issues/700, when we initialise an array, we first check if it should be a simple array or an arrayObject by checking the object type.

Further, when initialising variables in sub statements (for loops) which should be objects, we create arrayObjects.

Old output:
```php
        if ($object->isInitialized('portBindings') && null !== $object->getPortBindings()) {
            $values_10 = array();
            foreach ($object->getPortBindings() as $key => $value_10) {
                $values_11 = array();
                foreach ($value_10 as $value_11) {
                    $values_11[] = $this->normalizer->normalize($value_11, 'json', $context);
                }
                $values_10[$key] = $values_11;
            }
            $data['PortBindings'] = $values_10;
        }
```

New output:
```php
        if ($object->isInitialized('portBindings') && null !== $object->getPortBindings()) {
            $values_10 = new \ArrayObject(array(), \ArrayObject::ARRAY_AS_PROPS);
            foreach ($object->getPortBindings() as $key => $value_10) {
                $values_11 = array();
                foreach ($value_10 as $value_11) {
                    $values_11[] = new \ArrayObject($this->normalizer->normalize($value_11, 'json', $context), \ArrayObject::ARRAY_AS_PROPS);
                }
                $values_10[$key] = $values_11;
            }
            $data['PortBindings'] = $values_10;
        }
```

Rid commented:
It looks like the tests are also outputting incorrect code.

@Korbeil Can you take a look at the fix, if you think it looks good, I can update the tests.

Just an aside: I'm currently using the fix in production on the docker API and it's working well.

Korbeil reviewed (approved):
I didn't understood you wanted a review on the code before the commend you made today on the issue.
On the fly it seems totally ok to me, also I saw you already generated the `docker-php` client with and it seems it does fix your issue so I'm all for it :+1:

Rid commented:
Thanks @Korbeil, I'll fix the tests this or next week and then you can merge.

achton commented:
Thanks for this patch.

A bit of feedback: This patch fixes the case where Jane emits an empty array instead an empty object, but it introduces issues with other types, ie. nullable objects with local properties like:

```yaml
        caseAgent:
          type: object
          properties:
            name:
              type: string
              description: The name of the agent responsible for the case.
            phoneNumber:
              type: string
              description: A phone number for contacting the agent.
          required:
            - name
            - phoneNumber
          nullable: true
```

This changes the normalizer code to:

```php

$data['caseAgent'] = new \ArrayObject($this->normalizer->normalize($object->getCaseAgent(), 'json', $context), \ArrayObject::ARRAY_AS_PROPS);
```

and causes a fatal error when the value is `null`: `InvalidArgumentException: Passed variable is not an array or object in ArrayObject->__construct()`.

This may be the cause of the failing tests, in which case you can ignore me :-)

Korbeil commented:
Hey, any news about fixing tests ? :)

Rid commented:
@Korbeil I've fixed the tests, but it does not resolve @achton's issue.

We should check if the value is null at runtime and pass it as null instead of an object.

Rid commented:
Here's what it might look like if we test for null:

```php
    /**
     * @return array|string|int|float|bool|\ArrayObject|null
     */
    public function normalize($object, $format = null, array $context = array())
    {
        $data = array();
        if ($object->isInitialized('foo') && null !== $object->getFoo()) {
            $data['foo'] = $object->getFoo();
        }
        if ($object->isInitialized('bar') && null !== $object->getBar()) {
            $data['Bar'] = $object->getBar() == null ? null : new \ArrayObject($this->normalizer->normalize($object->getBar(), 'json', $context), \ArrayObject::ARRAY_AS_PROPS);
        }
        if ($object->isInitialized('baz') && null !== $object->getBaz()) {
            $data['Baz'] = $object->getBaz() == null ? null : new \ArrayObject($this->normalizer->normalize($object->getBaz(), 'json', $context), \ArrayObject::ARRAY_AS_PROPS);
        }
        foreach ($object as $key => $value) {
            if (preg_match('/.*/', (string) $key)) {
                $data[$key] = $value;
            }
        }
        return $data;
    }
```

It does a double check if we're using OpenAPIv3 and nullable is not true on the object and the object isn't null, it makes more sense for OpenAPIv2 and OpenAPIv3 where nullable is true, I think it does cover all schemas.

@Korbeil Let me know what you think?

Rid commented:
@Korbeil I've updated to handle the above and fixed the tests. If you agree with the changes I will squash the commits.

Korbeil commented:
Hey @Rid ,
Thanks for all the work on this PR !

You proposal for nullable handling seems totally fine to me.
The code seems fine to me :)

Korbeil commented:
Could you add a note in the CHANGELOG about this change ? :pray:

Rid commented:
@Korbeil I've updated the CHANGELOG and squashed the commits into one.

This closes:

https://github.com/janephp/janephp/issues/700
https://github.com/janephp/janephp/issues/680

Rid commented:
Ping @Korbeil

thePanz commented:
@Rid looks like you need to rebase and solve the conflicts on this PR

Rid commented:
@thePanz happy to rebase, do you have merge permissions?

thePanz commented:
> @thePanz happy to rebase, do you have merge permissions?

no I don't, but this PR is marked as not mergeable by github due to merge-conflicts

Rid commented:
@Korbeil are you happy to merge this? If so I'll work on a rebase.

GaetanColenoTiime commented:
Hi ! currently using Jane, it seems that i fall on this specific problem. 
Do we have an estimate on if and when this could be merged ? 
I'll try to use @Rid branch for now, so i could test it too !

Rid commented:
> Hi ! currently using Jane, it seems that i fall on this specific problem. Do we have an estimate on if and when this could be merged ? I'll try to use @Rid branch for now, so i could test it too !

It's up to @Korbeil, it was ready to merge almost a year and a half ago.

chrispenny commented:
I'm new to Jane PHP, but this was an issue that I came across immediately when I attempted a request to our API.

Our API accepts `null` or an empty object (when a request just wants to use defaults), and our Client was sending empty arrays (just as #700 describes), which is an invalid input for our API.

A workaround for us would have been if we could send `null` instead of the object/array, but the normalizers all seem to ignore setting fields when it detects that the value is `null`.

Thanks, @Rid for this fix (we're using it now). It would be great if this could be merged, as our Client really doesn't work without this.

Rid commented:
@chrispenny Please can you open a new issue referencing this PR? I don't think @Korbeil checks his notifications on PRs, maybe only issues?

Korbeil commented:
I do watch PR notification but I'm kinda against going back to ArrayObject for Objects.
ArrayObjects has many flaws and is not suited for this, I would prefer a custom made trait / object that we extend and does exactly what we do over it. We were before using it and it caused issues.

I tried doing something close to it there: https://github.com/Korbeil/jane-v8/blob/main/src/Component/JsonSchemaGenerator/Runtime/AdditionalAndPatternProperties.php
I do think this should be enough to handle what we are trying to make with ArrayObject without the usage of it.

Rid commented:
@Korbeil what type in PHP is suitable for creating empty objects? It's been a while since I looked at this, but last I remember was that only ArrayObject would result in an empty JSON object being outputted.

I can rewrite it with a custom trait but I'm struggling to think of what to use as the final empty object type.

Rid commented:
@Korbeil would you prefer to use a stdClass for empty objects or a custom lightweight object class?

Korbeil commented:
Like I said before, I would prefer a custom lightweight object class.

Korbeil commented:
Thanks @Rid for all the work on this! This was finally addressed in #998, but with a different approach than ArrayObject — as mentioned earlier in this thread, ArrayObject caused issues in the past. Jane now ships a custom lightweight `Runtime\JsonObject` class (`\stdClass`-based) in generated libraries, which serializes empty objects/maps as `{}` and handles nullable object properties safely.

Fixes the original goals of this PR (#700, #680). Closing as superseded by #998.

Rid commented:
Thanks for addressing this @Korbeil, unfortunately my time has been limited for contributions as of late.

We'll give this a test on https://github.com/beluga-php/docker-php-api
